File: leetcode/2100-2200/2106. 摘水果.cpp
//
// Created by 123 on 25-5-27.
//
#include "bits/stdc++.h"
using namespace std;
using ll = long long;
constexpr int MOD = 1'000'000'007;
constexpr int DIR[4][2] = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

class Solution {
public:
	int maxTotalFruits(vector<vector<int>>& fruits, int startPos, int k) {
		int n = fruits.size();
		vector<int> sum(n + 1);
		vector<int> indices(n);
		for (int i = 0; i < n; i++) {
			sum[i + 1] = sum[i] + fruits[i][1];
			indices[i] = fruits[i][0];
		}
		int ans = 0;
		for (int x = 0; x <= k / 2; x++) {
			/* 向左走 x 步，再向右走 k - x 步 */
			int y = k - 2 * x;
			int left = startPos - x;
			int right = startPos + y;
			int start = lower_bound(indices.begin(), indices.end(), left) - indices.begin();
			int end = upper_bound(indices.begin(), indices.end(), right) - indices.begin();
			ans = max(ans, sum[end] - sum[start]);
			/* 向右走 x 步，再向左走 k - x 步 */
			y = k - 2 * x;
			left = startPos - y;
			right = startPos + x;
			start = lower_bound(indices.begin(), indices.end(), left) - indices.begin();
			end = upper_bound(indices.begin(), indices.end(), right) - indices.begin();
			ans = max(ans, sum[end] - sum[start]);
		}
		return ans;
	}
};

class Solution {
public:
	int maxTotalFruits(vector<vector<int>> &fruits, int startPos, int k) {
		int left = lower_bound(fruits.begin(), fruits.end(), startPos - k, [](const auto &a, const auto b) {
			return a[0] < b;
		}) - fruits.begin();
		int right = left, s = 0, n = fruits.size();
		for (; right < n && fruits[right][0] <= startPos; right++) {
			s += fruits[right][1];
		}
		int ans = s;
		for (; right < n && fruits[right][0] <= startPos + k; ++right) {
			s += fruits[right][1]; // 枚举最右位置为 fruits[right][0]
			while (fruits[right][0] * 2 - fruits[left][0] - startPos > k &&
				   fruits[right][0] - fruits[left][0] * 2 + startPos > k)
				s -= fruits[left++][1]; // fruits[left][0] 无法到达
			ans = max(ans, s); // 更新答案最大值
		}
		return ans;
	}
};

class Solution2 {
public:
	int maxTotalFruits(vector<vector<int>>& fruits, int startPos, int k) {
		int n = fruits.size();
		vector<int> prefix_sum(500005, 0);

		// 构建前缀和数组
		for (auto& v : fruits) {
			prefix_sum[v[0]] += v[1];
		}
		for (int i = 1; i < 500005; i++) {
			prefix_sum[i] += prefix_sum[i - 1];
		}

		int ans = 0;

		// 策略1: 先向右走i步，再向左走j步（需要往返，所以总步数是i + i + j = 2*i + j）
		for (int i = 0; i <= k; i++) {  // 向右走i步
			int right = startPos + i;
			if (right >= 500005) break;

			int remaining = k - 2 * i;  // 剩余步数用于向左
			if (remaining < 0) break;

			int left = startPos - remaining;
			if (left < 0) left = 0;

			int current = prefix_sum[right] - (left > 0 ? prefix_sum[left - 1] : 0);
			ans = max(ans, current);
		}

		// 策略2: 先向左走i步，再向右走j步（需要往返，所以总步数是i + i + j = 2*i + j）
		for (int i = 0; i <= k; i++) {  // 向左走i步
			int left = startPos - i;
			if (left < 0) break;

			int remaining = k - 2 * i;  // 剩余步数用于向右
			if (remaining < 0) break;

			int right = startPos + remaining;
			if (right >= 500005) right = 500004;

			int current = prefix_sum[right] - (left > 0 ? prefix_sum[left - 1] : 0);
			ans = max(ans, current);
		}

		return ans;
	}
};

File: leetcode/2100-2200/2106. 摘水果.py
from collections import *
from functools import *
from math import *
from itertools import *
from heapq import *
from bisect import *
from operator import *
from typing import *
from queue import *
from string import *
from random import *

DIRS = [(-1, 0), (1, 0), (0, -1), (0, 1)]
Max = lambda a, b: b if b > a else a
Min = lambda a, b: b if b < a else a
INF = float("inf")
MOD = int(1e9 + 7)


class Solution:
    def maxTotalFruits(self, fruits: List[List[int]], startPos: int, k: int) -> int:
        # 找到最右边的位置
        if not fruits:
            return 0

        right_most = fruits[-1][0]
        left_most = fruits[0][0]

        # 扩展范围以包含可能的起始位置
        min_pos = min(left_most, startPos - k)
        max_pos = max(right_most, startPos + k)

        # 创建位置到水果数量的映射数组
        nums = [0] * (max_pos - min_pos + 1)
        for pos, count in fruits:
            nums[pos - min_pos] = count

        # 调整startPos到数组索引
        start_idx = startPos - min_pos

        ans = 0

        # 情况1：先往左走，再往右走
        # 初始化：只往左走k步
        left_bound = max(0, start_idx - k)
        window = sum(nums[left_bound : start_idx + 1])
        ans = max(ans, window)

        # 滑动窗口：逐步减少左边范围，增加右边范围
        right_steps = 1
        while left_bound + 2 <= start_idx and start_idx + right_steps < len(nums):
            # 往右多走一步
            window += nums[start_idx + right_steps]
            right_steps += 1

            # 左边减少两步的范围
            window -= nums[left_bound]
            left_bound += 1
            # if left_bound <= start_idx:
            window -= nums[left_bound]
            left_bound += 1

            ans = max(ans, window)

        # 情况2：先往右走，再往左走
        # 初始化：只往右走k步
        right_bound = min(len(nums) - 1, start_idx + k)
        window = sum(nums[start_idx : right_bound + 1])
        ans = max(ans, window)

        # 滑动窗口：逐步减少右边范围，增加左边范围
        left_steps = 1
        while right_bound - 2 >= start_idx and start_idx - left_steps >= 0:
            # 往左多走一步
            window += nums[start_idx - left_steps]
            left_steps += 1

            # 右边减少两步的范围
            window -= nums[right_bound]
            right_bound -= 1
            # if right_bound >= start_idx:
            window -= nums[right_bound]
            right_bound -= 1

            ans = max(ans, window)

        return ans
